### PR TITLE
Feat(#1238) : Stats profile activity

### LINKED ISSUE #1238: Show total amount of collected/sold items
When someone visits the profile, 
show 
- the total amount $KSM of sold items  and number of sold items
- how much the collector spent in lifetime in $KSM and the total count of items
- current holdings of the profile, total in $KSM and total items

When I will click on 
- Sold - **Sales** -- I will see how much the creator has sold items for in $KSM - count total sale events per NFT where transfer has happened
- Collected - **Holdings** -- Count value of current holdings, how much NFTs where is the current owner, invested (BUY) 
- Collected - **Total Invested** -- Total count, how much profile invested total in a lifetime, no matter if NFT is not owning anymore

Probably keep the same format as we have on collections under identical in the middle?

image

### Update
- Meanwhile, we have holdings per user https://beta.kodadot.xyz/rmrk/u/CdA62JpyfEyEASA5XKYJAyYZmdQPqe5X9x8MLnoTWtc9rNn?tab=holdings

### Ref
- #1539 
- #1285 
- #2822 

image

Could be reused to the #1165

### PR DESCRIPTION
**Pull request by MSghais**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [X] Feature
- [ ] Refactoring

### What's new?

- [X] PR closes #1238
- [X] Description : 
1. the total amount $KSM of sold items and number of sold items
2. how much the collector spent in lifetime in $KSM and the total count of items
3. current holdings of the profile, total in $KSM and total items

Good overview of profil

### Before submitting Pull Request, please make sure:

- [X] My contribution builds **clean without any errors or warnings**
- [] I've merged recent default branch -- **main** and I've no conflicts :
 conflict in package json and pnpm with new version deps.
Error.vue deleted otherwise nothing else.
I accept the Incoming changes when merge main into my branch.

- [X] I've tried to respect high code quality standards
- [X] I've didn't break any original functionality
- [X] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [X] I've tested PR on mobile and everything seems works
- [X] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [X] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=5DkJ2x6SCHsXw9jarxhv3mVx1cuGsWBLK7pZNnvdRmguDuyy

### Community participation

- [X] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [X] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

image

image

### CODE REVIEW

**Review by petersopko — CHANGES_REQUESTED** [review]
hey, thank you for your contribution :) can you please check if you see what I'm seeing here: 
https://deploy-preview-2995--koda-nuxt.netlify.app/rmrk/u/DSVZroU95KSVZKVr2pFdKfJkQtkgU5U1rkrBdeVnktHWmM4?tab=sold&page=1


Looks like a mismatch to me? 

image

**Review by MSghais — COMMENTED** [review]
All is self reviewed @petersopko.

For the mismatch you mentionned, I think this is related on other issue : https://github.com/kodadot/nft-gallery/issues/898.

**Review by MSghais — COMMENTED** [review]
All is self reviewed @petersopko.

For the mismatch you mentionned, I think this is related on other issue : https://github.com/kodadot/nft-gallery/issues/898.

**Comment by petersopko** [comment]
> All is self reviewed @petersopko.
> 
> For the mismatch you mentionned, I think this is related on other issue : #898.

It may be also related to the other issue, but if you look at the issue specs in #1238 (which you're trying to close by this PR), it seems that the task isn't done? 
Another example : https://deploy-preview-2995--koda-nuxt.netlify.app/rmrk/u/Gek2uCr3kWqdt8eSaYzE1GxNn3b1RB1FYgzyQuKhhJ63X7r?tab=sold
image

**Comment by MSghais** [comment]
> > All is self reviewed @petersopko.
> > For the mismatch you mentionned, I think this is related on other issue : #898.
> 
> It may be also related to the other issue, but if you look at the issue specs in #1238 (which you're trying to close by this PR), it seems that the task isn't done? Another example : https://deploy-preview-2995--koda-nuxt.netlify.app/rmrk/u/Gek2uCr3kWqdt8eSaYzE1GxNn3b1RB1FYgzyQuKhhJ63X7r?tab=sold image

Yes pretty strange issue. I receive the data after the query, but the check after apollo query see the data like a null values in this profil. 
Is just only on this profil, otherwise all profil work pretty well for me.

Bug - data

**Comment by petersopko** [comment]
@vikiival could you please take a look, not sure what to do with this one

File: components/rmrk/Profile/ProfilActivity.vue [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    })
+
+    this.totalSoldItems = soldEvents.length
+    const allValuesList = soldEvents.map((e) => parseFloat(e.meta))
+    const maxPriceSold = Math.max(...allValuesList)
+    // Highest Buy and Total amount sell
+    this.maxSoldPrice = maxPriceSold
+    this.totalSell = allValuesList
+      .map((x) => BigInt(x || 0))
+      .reduce((acc, cur) => acc + cur, BigInt(0))
```
Comment (vikiival) [inline components/rmrk/Profile/ProfilActivity.vue:193]: DRY

File: components/rmrk/Profile/ProfileDetail.vue [hunk 2]
```diff
@@ -34,6 +34,9 @@
           <ProfileLink :address="id" :inline="true" showTwitter showDiscord />
         </div>
       </div>
+      <div class="column is-6-tablet is-7-desktop is-8-widescreen">
+        <ProfilActivity :id="id" />
```
Comment (vikiival) [inline components/rmrk/Profile/ProfileDetail.vue:38]:
```suggestion
        <ProfileActivity :id="id" />
```

File: langDir/en.json [hunk 3]
```diff
@@ -566,6 +566,19 @@
     "distribution": "DISTRIBUTION",
     "floorPrice": "FLOOR PRICE"
   },
+  "profil_stats": {
```
Comment (vikiival) [inline langDir/en.json:569]:
```suggestion
  "profileStats": {
```

File: queries/rmrk/subsquid/profilStatsById.graphql [hunk 4]
```diff
@@ -0,0 +1,99 @@
+query profilStatsById($id: String!, $search_listed: [EventWhereInput!],  $search_collected: [NFTEntityWhereInput!],  $search_sold: [EventWhereInput!], $search_invested:[EventWhereInput!]) {
```
Comment (vikiival) [inline queries/rmrk/subsquid/profilStatsById.graphql:1]:
```suggestion
query profileStatsById($id: String!, $search_listed: [EventWhereInput!],  $search_collected: [NFTEntityWhereInput!],  $search_sold: [EventWhereInput!], $search_invested:[EventWhereInput!]) {
```

File: queries/rmrk/subsquid/profilStatsById.graphql [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  listed : events(where: {  caller_eq: $id , AND: $search_listed },  ) {
```
Comment (vikiival) [inline queries/rmrk/subsquid/profilStatsById.graphql:2]: pleease use fragments sir

File: components/rmrk/Profile/ProfilActivity.vue [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    const { data } = await this.$apollo.query({
+      query: profilStatsById,
+      client: 'subsquid',
+      variables: {
+        id: this.id,
+        search_sold: {
+          interaction_eq: 'SEND',
+          nft: { name_not_contains: '%Kanaria%', burned_eq: false },
+        },
+        search_listed: {
```
Comment (vikiival) [inline components/rmrk/Profile/ProfilActivity.vue:123]: Please follow code guidline :)

**Review by vikiival — CHANGES_REQUESTED** [review]

**Review by MSghais — COMMENTED** [review]
Feedback done, clean & refacto

File: queries/rmrk/subsquid/profileStatsById.graphql [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+          id
+          blockNumber
+          timestamp
+          caller
+          currentOwner
+          meta
+          interaction
+      }
+    }
+  }
```
Comment (vikiival) [inline queries/rmrk/subsquid/profileStatsById.graphql:55]: wtf

Comment (vikiival) [reply queries/rmrk/subsquid/profileStatsById.graphql:55]: You query events to get nfts to get events?

File: components/rmrk/Profile/ProfileActivity.vue [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+      <div class="level-item has-text-centered">
+        <div>
+          <p class="title">
+            {{ totalPurchases }}
+          </p>
+          <p class="heading">
+            {{ $t('profileStats.totalBuys') }}
+          </p>
+        </div>
+      </div>
```
Comment (vikiival) [inline components/rmrk/Profile/ProfileActivity.vue:31]: Make a component

File: components/rmrk/Profile/ProfileActivity.vue [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+            name_not_contains: '%Kanaria%',
+            burned_eq: false,
+          },
+        },
+        search_collected: { currentOwner_eq: this.id, burned_eq: false },
+
+        search_invested: {
+          interaction_eq: 'BUY',
+          nft: { name_not_contains: '%Kanaria%', burned_eq: false },
+        },
```
Comment (vikiival) [inline components/rmrk/Profile/ProfileActivity.vue:137]: This should be handled by query

File: components/rmrk/Profile/ProfileActivity.vue [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    data.sold.forEach((e: Event) => {
+      if (e.nft && e.nft.events && e.nft.events.length > 0) {
+        const buyEvents = pairListBuyEvent(
+          e.nft.events as unknown as Interaction[]
+        )
+        buyEvents.forEach((ev) => {
+          soldEvents.push(ev)
+        })
+      }
+    })
```
Comment (vikiival) [inline components/rmrk/Profile/ProfileActivity.vue:215]: what does this do?

**Review by vikiival — CHANGES_REQUESTED** [review]

File: components/rmrk/Profile/ProfileActivity.vue (hunk 10 above)
Comment (vikiival) [reply components/rmrk/Profile/ProfileActivity.vue:194]: rubick 006 contains current owner for the event

File: components/shared/ScrollTopButton.vue [hunk 11]
```diff
@@ -0,0 +1,50 @@
+<template>
```
Comment (vikiival) [inline components/shared/ScrollTopButton.vue:1]: ?

File: components/shared/format/StatsColumn.vue [hunk 12]
```diff
@@ -0,0 +1,65 @@
+<template>
+  <div class="level-item has-text-centered">
+    <div>
+      <p class="title">{{ value }}</p>
+      <p class="heading">
+        {{ $t(header) }}
+      </p>
+    </div>
+  </div>
+</template>
```
Comment (vikiival) [inline components/shared/format/StatsColumn.vue:10]:
```suggestion
<template>
      <div class="level-item has-text-centered">
        <div>
          <p class="title">
             <slot name="title"></slot>
          </p>
          <p class="heading">
             <slot name="heading"></slot>
          </p>
        </div>
      </div>
</template>
```

**Review by vikiival — CHANGES_REQUESTED** [review]

**Comment by vikiival** [comment]
```graphql
query profileStatsByAccount($id: String!) {
  listed: nftEntitiesConnection(where: {currentOwner_eq: $id, price_gt: "0"}, orderBy: name_ASC) {
    totalCount
  }
  collected: nftEntitiesConnection(where: {currentOwner_eq: $id, issuer_not_eq: $id, events_some: {interaction_eq: BUY}}, orderBy: name_ASC) {
    totalCount
    edges {
      node {
        events(where: {interaction_eq: BUY, caller_eq: $id}) {
          meta
        }
      }
    }
  }
  sold: nftEntitiesConnection(where: {currentOwner_not_eq: $id, issuer_eq: $id, events_some: {interaction_eq: BUY}}, orderBy: name_ASC) {
    totalCount
    edges {
      node {
        events(where: {interaction_eq: BUY, currentOwner_eq: $id}) {
          meta
        }
      }
    }
  }
  obtained: nftEntitiesConnection(where: {currentOwner_eq: $id, issuer_not_eq: $id}, orderBy: name_ASC) {
    totalCount
  }
}

```

let me know if you are missing any value

**Review by MSghais — COMMENTED** [review]
Add new subquery more fast

**Review by MSghais — COMMENTED** [review]
Done

File: components/shared/format/StatsColumn.vue [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+<script lang="ts">
+import { Component, Prop, Vue } from 'nuxt-property-decorator'
+import { checkInvalidBalanceFilter } from '@/utils/formatBalance'
+@Component({
+  filters: {
+    checkInvalidBalance: checkInvalidBalanceFilter,
+    round: function roundOffNumber(value, limit) {
```
Comment (vikiival) [inline components/shared/format/StatsColumn.vue:18]: move to utils

File: components/shared/format/StatsColumn.vue [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    </div>
+  </div>
+</template>
+
+<script lang="ts">
+import { Component, Prop, Vue } from 'nuxt-property-decorator'
+import { checkInvalidBalanceFilter } from '@/utils/formatBalance'
+@Component({
+  filters: {
+    checkInvalidBalance: checkInvalidBalanceFilter,
```
Comment (vikiival) [inline components/shared/format/StatsColumn.vue:17]:
```suggestion
import { checkInvalidBalanceFilter as checkInvalidBalance } from '@/utils/formatBalance'

@Component({
  filters: {
    checkInvalidBalance,
```

File: components/shared/format/StatsColumn.vue [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        maximumFractionDigits: limit,
+      })
+    },
+  },
+})
+export default class StatsColumn extends Vue {
+  @Prop({ default: 0 }) readonly value: number | string | undefined
+  @Prop({ default: 'profileStats.totalBuys' }) readonly header:
+    | string
+    | undefined
```
Comment (vikiival) [inline components/shared/format/StatsColumn.vue:32]:
```suggestion
  @Prop({ type: String, default: 'profileStats.totalBuys' }) readonly label: string
```

File: components/shared/format/StatsColumn.vue [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  @Prop(Boolean) readonly inline!: boolean
+  @Prop(Boolean) readonly hideUnit!: boolean
+
+  private readonly coinId: string = 'kusama'
```
Comment (vikiival) [inline components/shared/format/StatsColumn.vue:36]: What's this?

File: components/shared/format/StatsColumn.vue [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    return this.$store.getters['chain/getChainProperties']
+  }
+
+  get decimals(): number {
+    return this.chainProperties.tokenDecimals
+  }
+
+  get unit(): string {
+    return this.chainProperties.tokenSymbol
+  }
```
Comment (vikiival) [inline components/shared/format/StatsColumn.vue:49]: Use chainMixin

**Review by vikiival — CHANGES_REQUESTED** [review]
Some stuff

**Review by MSghais — COMMENTED** [review]
Stuff done

**Comment by vikiival** [comment]
Screenshot 2022-05-23 at 15 52 30

**Comment by vikiival** [comment]
@petersopko can u pls check? 🥺

**Comment by petersopko** [comment]
> hey, thank you for your contribution :) can you please check if you see what I'm seeing here: https://deploy-preview-2995--koda-nuxt.netlify.app/rmrk/u/DSVZroU95KSVZKVr2pFdKfJkQtkgU5U1rkrBdeVnktHWmM4?tab=sold&page=1
> 
> Looks like a mismatch to me?
> 
> image

@MSghais your added feature still doesn't work. Mismatch between sold in the tabs vs. your added feature is still there.
Currently, the totalCount of sold items should be 300. Please check again. This is what the added query shows:
image

**Comment by petersopko** [comment]
> > > hey, thank you for your contribution :) can you please check if you see what I'm seeing here: https://deploy-preview-2995--koda-nuxt.netlify.app/rmrk/u/DSVZroU95KSVZKVr2pFdKfJkQtkgU5U1rkrBdeVnktHWmM4?tab=sold&page=1
> > > Looks like a mismatch to me?
> > > image
> > 
> > 
> > @MSghais your added feature still doesn't work. Mismatch between sold in the tabs vs. your added feature is still there. Currently, the totalCount of sold items should be 300. Please check again. This is what the added query shows: image
> 
> @petersopko for the mismatch you mentionned, this is related on other issue : #898.
> 
> In my query, I take all Sold event. In the Tab you compare, we have also the Gifted event related on the issue above, that's why the difference is here for the totalCount.


Okay, but the number still isn't correct? Another example, when I check the query below at rubick https://app.gc.subsquid.io/beta/rubick/006/graphql with account `CgGKH1dsPgJkQ3caFsnDtMUWUqVJKEpRt1ekM5RDW4UrYhP` I see total count of 66. I think the query is correct. When I check the profile stats there, I see the number 50. This has nothing to do with the tab. Also, sometimes this happens: 
image 

Data just doesn't load?


```graphql 
sold: nftEntitiesConnection(
    where: {
      currentOwner_not_eq: $id
      issuer_eq: $id
      name_not_eq: "%Kanaria%"
      burned_not_eq: true
      events_some: { interaction_eq: BUY }
    }
    orderBy: name_ASC
  ) {
    totalCount
    edges {
      node {
        events(
          where: {
            interaction_eq: BUY
            currentOwner_eq: $id
            caller_not_eq: $id
          }
        ) {
          meta
        }
      }
    }
  }
```

**Review by petersopko — APPROVED** [review]
lgtm

**Review by vikiival — APPROVED** [review]

**Comment by petersopko** [comment]
pay 200 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $200 @ 71.09 USD/KSM ~ 2.813 $KSM
🧗 [5DkJ2x6SCHsXw9jarxhv3mVx1cuGsWBLK7pZNnvdRmguDuyy](https://kusama.subscan.io/account/5DkJ2x6SCHsXw9jarxhv3mVx1cuGsWBLK7pZNnvdRmguDuyy)
🔗 [0x01896f925b34b70a39f9bd3fc5761414a01034bc99d2d31323a1ddb62612e8ee](https://kusama.subscan.io/extrinsic/0x01896f925b34b70a39f9bd3fc5761414a01034bc99d2d31323a1ddb62612e8ee)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**